// protoc zksnark.proto --go_out=plugins=grpc:zksnark
// protoc zksnark.proto --cpp_out=.
// protoc zksnark.proto --grpc_out=. --plugin=protoc-gen-grpc=`which grpc_cpp_plugin`
syntax = "proto3";

package zksnark;

service Zksnark {
    rpc Prove(ProveRequest) returns (ProveReply) {}
    rpc Verify(VerifyRequest) returns (VerifyReply) {}
}

message MerkleHash {
    bytes hash = 1;
}

message MerklePath {
    repeated MerkleHash authPath = 1;
    repeated bool index = 2;
}

message Note {
    uint64 value = 1;
    bytes cm = 2;  // 32 bytes
    bytes r = 3;
    bytes rho = 4;
    bytes apk = 5;
    bytes nf = 6;
}

message JSInput {
    MerklePath witnessPath = 1;
    bytes spendingKey = 2;
    Note note = 3;
}

message ProveRequest {
    repeated JSInput inputs = 1;
    repeated Note outNotes = 2;
    bytes hsig = 3;
    bytes phi = 4;
    repeated bytes rts = 5;
    uint64 reward = 6;
    uint64 fee = 7;
    uint32 addressLastByte = 8;
}

message PHGRProof {
    bytes g_A = 1;
    bytes g_A_prime = 2;
    bytes g_B = 3;
    bytes g_B_prime = 4;
    bytes g_C = 5;
    bytes g_C_prime = 6;
    bytes g_K = 7;
    bytes g_H = 8;
}

message ProveReply {
    PHGRProof proof = 1;
    bool success = 2;
}

message VerifyRequest {
    PHGRProof proof = 1;
    bytes hsig = 2;
    repeated bytes rts = 3;
    repeated bytes nullifiers = 4;
    repeated bytes commits = 5;
    repeated bytes macs = 6;
    uint64 reward = 7;
    uint64 fee = 8;
    uint32 addressLastByte = 9;
}

message VerifyReply {
    bool valid = 1;
    bool success = 2;
}
